rpc/rpcserver/server.go

```diff
@@ -2314,3 +2328,125 @@ func (s *messageVerificationServer) VerifyMessage(ctx context.Context, req *pb.V
 WrongAddrKind:
 	return nil, status.Error(codes.InvalidArgument, "address must be secp256k1 P2PK or P2PKH")
 }
+
+// StartDecodeMessageService starts the MessageDecode service
+func StartDecodeMessageService(server *grpc.Server, wallet *wallet.Wallet) {
+	decodeMessageService.chainParams = wallet.ChainParams()
+}
+
+func marshalDecodedTxInputs(mtx *wire.MsgTx) []*pb.DecodedTransaction_Input {
+	inputs := make([]*pb.DecodedTransaction_Input, len(mtx.TxIn))
+
+	for i, txIn := range mtx.TxIn {
+		// The disassembled string will contain [error] inline
+		// if the script doesn't fully parse, so ignore the
+		// error here.
+		disbuf, _ := txscript.DisasmString(txIn.SignatureScript)
+
+		inputs[i] = &pb.DecodedTransaction_Input{
+			PreviousTransactionHash:  txIn.PreviousOutPoint.Hash[:],
+			PreviousTransactionIndex: txIn.PreviousOutPoint.Index,
+			Tree:               pb.DecodedTransaction_Input_TreeType(txIn.PreviousOutPoint.Tree),
+			Sequence:           txIn.Sequence,
+			AmountIn:           txIn.ValueIn,
+			BlockHeight:        txIn.BlockHeight,
+			BlockIndex:         txIn.BlockIndex,
+			SignatureScript:    txIn.SignatureScript[:],
+			SignatureScriptAsm: disbuf,
+		}
+	}
+
+	return inputs
+}
+
+func marshalDecodedTxOutputs(mtx *wire.MsgTx, chainParams *chaincfg.Params) []*pb.DecodedTransaction_Output {
+	outputs := make([]*pb.DecodedTransaction_Output, len(mtx.TxOut))
+	txType := stake.DetermineTxType(mtx)
+
+	for i, v := range mtx.TxOut {
+		// The disassembled string will contain [error] inline if the
+		// script doesn't fully parse, so ignore the error here.
+		disbuf, _ := txscript.DisasmString(v.PkScript)
+
+		// Attempt to extract addresses from the public key script.  In
+		// the case of stake submission transactions, the odd outputs
+		// contain a commitment address, so detect that case
+		// accordingly.
+		var addrs []dcrutil.Address
+		var encodedAddrs []string
+		var scriptClass txscript.ScriptClass
+		var reqSigs int
+		var commitAmt *dcrutil.Amount
+		if txType == stake.TxTypeSStx && (i%2 != 0) {
```

Comment:
this logic is getting a bit more complicated then i'd like to see in the rpc service implementation.  consider adding this to the wallet or stake packages